internal/lsp/documentsymbol.go

```diff
@@ -0,0 +1,267 @@
+			ruleSymbol := types.DocumentSymbol{
+				Name:           refToString(rule.Head.Ref()),
+				Kind:           kind,
+				Range:          ruleRange,
+				SelectionRange: ruleRange,
+			}
+
+			if detail := getRuleDetail(rule); detail != "" {
+				ruleSymbol.Detail = &detail
+			}
+
+			pkgSymbols = append(pkgSymbols, ruleSymbol)
+		} else {
+			groupFirstRange := locationToRange(rules[0].Location)
+			groupLastRange := locationToRange(rules[len(rules)-1].Location)
+
+			groupRange := types.Range{
+				Start: groupFirstRange.Start,
+				End:   groupLastRange.End,
+			}
+
+			kind := symbols.Variable
+			if rules[0].Head.Args != nil {
+				kind = symbols.Function
+			}
+
+			groupSymbol := types.DocumentSymbol{
+				Name:           refToString(rules[0].Head.Ref()),
+				Kind:           kind,
+				Range:          groupRange,
+				SelectionRange: groupRange,
+			}
+
+			detail := getRuleDetail(rules[0])
+			if detail != "" {
+				groupSymbol.Detail = &detail
+			}
+
+			children := make([]types.DocumentSymbol, 0, len(rules))
+
+			for i, rule := range rules {
+				childRange := locationToRange(rule.Location)
+				childRule := types.DocumentSymbol{
+					Name:           fmt.Sprintf("#%d", i+1),
+					Kind:           kind,
+					Range:          childRange,
+					SelectionRange: childRange,
+				}
+
+				childDetail := getRuleDetail(rule)
+				if childDetail != "" {
+					childRule.Detail = &childDetail
+				}
+
+				children = append(children, childRule)
+			}
+
+			groupSymbol.Children = &children
+
+			pkgSymbols = append(pkgSymbols, groupSymbol)
+		}
+	}
+
+	if len(pkgSymbols) > 0 {
+		pkg.Children = &pkgSymbols
+	}
+
+	docSymbols = append(docSymbols, pkg)
+
+	return docSymbols
+}
+
+func locationToRange(location *ast.Location) types.Range {
+	lines := bytes.Split(location.Text, []byte("\n"))
+
+	var endLine uint
+	if len(lines) == 1 {
+		endLine = uint(location.Row - 1)
+	} else {
+		endLine = uint(location.Row-1) + uint(len(lines)-1)
+	}
+
+	return types.Range{
+		Start: types.Position{
+			Line:      uint(location.Row - 1),
+			Character: uint(location.Col - 1),
+		},
+		End: types.Position{
+			Line:      endLine,
+			Character: uint(len(lines[len(lines)-1])),
+		},
+	}
+}
+
+func refToString(ref ast.Ref) string {
+	sb := strings.Builder{}
+
+	for i, part := range ref {
+		if part.IsGround() {
+			if i > 0 {
+				sb.WriteString(".")
+			}
+
+			sb.WriteString(strings.Trim(part.Value.String(), `"`))
+		} else {
+			if i == 0 {
+				sb.WriteString(strings.Trim(part.Value.String(), `"`))
+			} else {
+				sb.WriteString("[")
+				sb.WriteString(strings.Trim(part.Value.String(), `"`))
+				sb.WriteString("]")
+			}
+		}
+	}
+
+	return sb.String()
+}
+
+//nolint:nestif
+func getRuleDetail(rule *ast.Rule) string {
+	if rule.Head.Args != nil {
+		return "function" + rule.Head.Args.String()
+	}
+
+	if rule.Head.Key != nil && rule.Head.Value == nil {
+		return "multi-value rule"
+	}
+
+	if rule.Head.Value != nil {
```

Comment:
could exit early here for `rule.Head.Value == nil` and then avoid a level of nesting in the latter half? Just an idea, feel free to ignore. If `getRuleDetail` is going to be extended for lots of other cases it might not be a good idea.